### ISSUE TITLE
[doc] opening nn page is very slow

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#8529]
For example, see https://pytorch.org/docs/stable/nn.html#torch.nn.functional.normalize

Two reasons:

1. It renders a **lot** of mathjax.
2. We put both modules and functionals on the same html.

Can we speed it up possibly with some refactoring?

**Comment by soumith** [comment]
the fix here is to move from mathJax to Katex, which is MUCH faster: https://khan.github.io/KaTeX/

**Comment by soumith** [comment]
@goodlux said he'd look into it

**Comment by goodlux** [comment]
I was able to get Katex running relatively easily. It just involves replacing the MathJax extension with Katex, and pointing the conf.py file to the css and javascript files. 

Many things render well, but It doesn't seem to handle multi-line equations, like the ones beginning with:  \begin{equation*}

**Comment by SsnL** [comment]
@goodlux A quick look at their doc make me think that they might be just assuming math env. Try removing the `equation*` env wrapper?

**Comment by goodlux** [comment]
I just tried, and this worked for the equation in question. So that's a plus. 

There still seems to be an issue in some other spots:
For instance, the output on the data shape output for 3dConv is getting mangled.

**Comment by SsnL** [comment]
@goodlux Thanks. I also think that 2 is better. For the above example, my guess is that it is a multiline issue causing \left to not matched with the \right on the next line. Probably just line continuation would work, i.e. 
```
      .. math::
          D_{out} = \left\lfloor\frac{D_{in} + 2 * \text{padding}[0] - \text{dilation}[0] \
                * (\text{kernel_size}[0] - 1) - 1}{\text{stride}[0]} + 1\right\rfloor

          H_{out} = \left\lfloor\frac{H_{in} + 2 * \text{padding}[1] - \text{dilation}[1] \
                * (\text{kernel_size}[1] - 1) - 1}{\text{stride}[1]} + 1\right\rfloor

          W_{out} = \left\lfloor\frac{W_{in} + 2 * \text{padding}[2] - \text{dilation}[2] \
                * (\text{kernel_size}[2] - 1) - 1}{\text{stride}[2]} + 1\right\rfloor
```

Could you try it?

**Comment by goodlux** [comment]
Ok, I had a minute to debug this. It turns out that the \left and \right don't seem to make any visual difference in Katex, and also they are expressed "\left." "\right.", with a period. 

So I did this:

.. math::
    D_{out} = 
    \lfloor
    \frac
    {D_{in} + 2 * {padding}[0] - {dilation}[0] 
      * ({kernel\_size}[0] - 1) - 1} 
    {{stride}[0]} + 1
    \rfloor

Which renders:

<img width="593" alt="screen shot 2018-06-20 at 12 02 18 pm" src="https://user-images.githubusercontent.com/31841/41679293-88cb68bc-7482-11e8-9ef4-8b4eb44c7932.png">

VS. the original:

<img width="662" alt="screen shot 2018-06-20 at 12 02 57 pm" src="https://user-images.githubusercontent.com/31841/41679296-903b434c-7482-11e8-8a15-63be0818625e.png">

What do you think? Is this a suitable replacement? The floor delimiters are a bit smaller, but I think this conveys the same meaning. 

---

In any case, it is going to be a bit of work to debug and proof these equations, both for existing content, and new content that may have been produced with MathJax. I don't mind going through these and fixing, but I want to make sure everyone involved is OK with the transition.

**Comment by SsnL** [comment]
@goodlux This seems weird as the Katex doc claims them to be working: https://khan.github.io/KaTeX/function-support.html (Delimiter Sizing section).

**Comment by SsnL** [comment]
@goodlux I just tried, and it seems that KaTeX requires manually escaping underscore even in \text. Can you try that, i.e., using `\text{kernel_size}`

**Comment by asprasan** [comment]
Why don't we put each function in a separate html page just like numpy/scipy (or even tensorflow) does instead of listing everything in a single page? Instead of spending time on optimizing the equations this could be a much better alternative. I guess, most people visit the docs page only to check out one or two functions.

**Comment by goodlux** [comment]
@asprasan Agreed, that is good practice, and we are going out your suggestion in addition to rolling out Katex, in a separate PR.
